include/lbann/weights/weights_proxy.hpp

```diff
@@ -0,0 +1,398 @@
+   *  for WeightsProxy objects of the same static type.
+   */
+  WeightsProxy(WeightsProxy&& other) noexcept
+    : master_weights_{other.master_weights_},
+      values_{std::move(other.values_)}
+  {
+    other.clear();
+  }
+
+  /** @brief Destructor */
+  ~WeightsProxy() noexcept
+  {
+    this->clear();
+  }
+
+  ///@}
+
+  /** @name Assignment operators */
+  ///@{
+
+  /** @brief Copy assignment operator. */
+  WeightsProxy& operator=(WeightsProxy const& other)
+  {
+    WeightsProxy<TensorDataType>(other).swap(*this);
+    return *this;
+  }
+
+  /** @brief Assignment from WeightsProxy object of a different type.
+   *
+   *  After assignment, @c this and @c other both proxy the weights
+   *  proxied by @c other.
+   *
+   *  @tparam T (Deduced) The type of the input weights object's
+   *                      values.
+   */
+  template <typename T>
+  WeightsProxy& operator=(WeightsProxy<T> const& other)
+  {
+    WeightsProxy<TensorDataType>(other).swap(*this);
+    return *this;
+  }
+
+  /** @brief Assignment from data_type_weights object.
+   *
+   *  @tparam T (Deduced) The type of the input weights object's
+   *                      values.
+   */
+  template <typename T>
+  WeightsProxy& operator=(data_type_weights<T> const& w)
+  {
+    this->setup(w);
+    return *this;
+  }
+
+  /** @brief Assignment from data_type_weights object */
+  WeightsProxy& operator=(weights const& w)
+  {
+    this->setup(w);
+    return *this;
+  }
+
+  /** @brief Move assignment from another proxy object. */
+  WeightsProxy& operator=(WeightsProxy&& other) noexcept
+  {
+    // "Move-and-swap" idiom
+    WeightsProxy<TensorDataType>(std::move(other)).swap(*this);
+    return *this;
+  }
+
+  ///@}
+  /** @name Master object management and synchronization. */
+  ///@{
+
+  /** @brief Restore the default state of the proxy.
+   *
+   *  After this function is called, the object will be empty().
+   */
+  void clear() noexcept
+  {
+    values_.reset();
+    master_weights_ = nullptr;
+  }
+
+  /** @brief Provide setup function for delayed construction.
+   *
+   *  This overwrites any existing data.
+   *
+   *  @param w The weights object to be proxied.
+   */
+  void setup(weights const& w)
+  {
+    this->internal_setup_(w);
+  }
+
+  /** @brief Provide setup function for delayed construction.
+   *
+   *  This overwrites any existing data.
+   *
+   *  @param w The weights object to be proxied.
+   *
+   *  @tparam T (Deduced) The type of the input weights object's
+   *                      values.
+   */
+  template <typename T>
+  void setup(data_type_weights<T> const& w)
+  {
+    this->internal_setup_(w);
+  }
+
+  /** @brief Synchronize the held values with the master set.
+   *
+   *  If empty(), this function takes the view that there is no master
+   *  with which to synchronize, so no action is required -- it is a
+   *  no-op.
+   */
+  void synchronize_with_master()
```

Comment:
If multiple layers have a proxy to a weight, this synchronize would clobber each other, right?  Is there a synch collective allreduce operations?